src/db.rs

```diff
@@ -260,43 +261,60 @@ impl DbInner {
-            // iterate over the WAL SSTs in reverse order to ensure we recover in write-order
-            // buffer the WAL entries to bulk replay them into the memtable.
-            let mut wal_replay_buf = Vec::new();
-            while let Some(kv) = iter.next_entry().await? {
-                wal_replay_buf.push(kv);
+
+        for sst_chunk in wal_sst_list.chunks(4) {
+            let mut wal_tasks = Vec::new();
+            let mut wal_replay_errors = Vec::new();
+            // constructs a list of tokio tasks to asynchronously fetch wal sst
+            for &sst_id in sst_chunk {
+                let this = self.clone();
+                wal_tasks.push(tokio::spawn(async move { this.fetch_wal(sst_id).await }));
             }
-            {
-                let mut guard = self.state.write();
-                for kv in wal_replay_buf.iter() {
-                    match &kv.value {
-                        ValueDeletable::Value(value) => {
-                            guard.memtable().put(kv.key.as_ref(), value.as_ref())
+
+            // executes all the `fetch_wal` tasks in parallel
+            let results = join_all(wal_tasks).await;
+
+            // iterate over ordered wal_replay_buffers and build the memtable
+            for result in results {
+                match result.expect("join task for wal replay failed") {
+                    Ok(r) => {
+                        let sst_id = r.0;
+                        let wal_replay_buf = r.1;
+                        last_sst_id = sst_id;
+                        {
+                            let mut guard = self.state.write();
+                            for kv in wal_replay_buf.iter() {
+                                match &kv.value {
+                                    ValueDeletable::Value(value) => {
+                                        guard.memtable().put(kv.key.as_ref(), value.as_ref())
+                                    }
+                                    ValueDeletable::Tombstone => {
+                                        guard.memtable().delete(kv.key.as_ref())
+                                    }
+                                }
+                            }
+                            self.maybe_freeze_memtable(&mut guard, sst_id);
+                            if guard.state().core.next_wal_sst_id == sst_id {
+                                guard.increment_next_wal_id();
+                            }
                         }
-                        ValueDeletable::Tombstone => guard.memtable().delete(kv.key.as_ref()),
                     }
+                    Err(e) => wal_replay_errors.push(e),
                 }
-                self.maybe_freeze_memtable(&mut guard, sst_id);
-                if guard.state().core.next_wal_sst_id == sst_id {
-                    guard.increment_next_wal_id();
+
+                if !wal_replay_errors.is_empty() {
+                    return Err(SlateDBError::WalReplayError(wal_replay_errors));
```

Comment:
I don't think we need `WalReplayError` here. `fetch_wal` already returns a `SlateDBError`, which we can return here rather than wrapping it in another error.